File: cpp/euklides_optymalny.cpp
/*
 * euklides.cpp
 */

#include <iostream>

using namespace std;

int nwd(int a, int b);

int nwd_rek(int a, int b)
{
    if(b==0)
    {
        return a;
    }
    return nwd_rek(b, a%b);
}

int main(int argc, char **argv)
{
    int a = 0;
    int b = 0;
    
    cout<<"Podaj pierwsza liczbe: ";
    cin>>a;
    cout<<"Podaj druga liczbe: ";
    cin>>b;
    
    cout<<"NWD: "<<nwd(a,b)<<endl;
    cout<<"NWD_rek: "<<nwd_rek(a,b);
    return 0;
}

int nwd(int a, int b)
{
    while(a > 0)
    {
        a = a%b;
        b-=a;
    }
    return b;
}

File: python/euklides_optymalny.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-


def nwd(a, b):
    while a > 0:
        a = a % b
        b = b - a
    return b


def nwd_rek(a, b):
    if b == 0:
        return a
    return nwd_rek(b, a % b)


def main(args):
    a = int(input("Podaj 1 liczbe: "))
    b = int(input("Podaj 2 liczbe: "))

    assert nwd(1989, 867) == 51

    # print("Najwiekszy wspolny dzielnik: ", nwd(a,b))
    # print("Nwd({:d},{:d}) = {:d}".format(a, b, nwd(a, b)))
    print("Nwd({:d},{:d}) = {:d}".format(a, b, nwd_rek(a, b)))

    return 0


if __name__ == '__main__':
    import sys
    sys.exit(main(sys.argv))
